libraries/chain/include/eosio/chain/log_catalog.hpp

```diff
@@ -0,0 +1,283 @@
+         auto path_without_extension = log_path.parent_path() / log_path.stem().string();
+
+         LogData log(log_path);
+
+         verifier.verify(log, log_path);
+
+         // check if index file matches the log file
+         if (!index_matches_data(index_path, log))
+            log.construct_index(index_path);
+
+         auto existing_itr = collection.find(log.first_block_num());
+         if (existing_itr != collection.end()) {
+            if (log.last_block_num() <= existing_itr->second.last_block_num) {
+               wlog("${log_path} contains the overlapping range with ${existing_path}.log, dropping ${log_path} "
+                    "from catalog",
+                    ("log_path", log_path.string())("existing_path", existing_itr->second.filename_base.string()));
+               return;
+            } else {
+               wlog(
+                   "${log_path} contains the overlapping range with ${existing_path}.log, droping ${existing_path}.log "
+                   "from catelog",
+                   ("log_path", log_path.string())("existing_path", existing_itr->second.filename_base.string()));
+            }
+         }
+
+         collection.insert_or_assign(log.first_block_num(), mapped_type{log.last_block_num(), path_without_extension});
+      });
+   }
+
+   bool index_matches_data(const bfs::path& index_path, const LogData& log) const {
+      if (!bfs::exists(index_path))
+         return false;
+
+      auto num_blocks_in_index = bfs::file_size(index_path) / sizeof(uint64_t);
+      if (num_blocks_in_index != log.num_blocks())
+         return false;
+
+      // make sure the last 8 bytes of index and log matches
+      fc::cfile index_file;
+      index_file.set_file_path(index_path);
+      index_file.open("r");
+      index_file.seek_end(-sizeof(uint64_t));
+      uint64_t pos;
+      index_file.read(reinterpret_cast<char*>(&pos), sizeof(pos));
+      return pos == log.last_block_position();
+   }
+
+   std::optional<uint64_t> get_block_position(uint32_t block_num, mapmode mode = mapmode::readonly) {
+      try {
+         if (active_index != npos) {
+            auto active_item = collection.nth(active_index);
+            if (active_item->first <= block_num && block_num <= active_item->second.last_block_num &&
+                log_data.flags() == mode) {
+               return log_index.nth_block_position(block_num - log_data.first_block_num());
+            }
+         }
+         if (collection.empty() || block_num < collection.begin()->first)
+            return {};
+
+         auto it = --collection.upper_bound(block_num);
+
+         if (block_num <= it->second.last_block_num) {
+            auto name = it->second.filename_base;
+            log_data.open(name.replace_extension("log"), mode);
+            log_index.open(name.replace_extension("index"));
+            this->active_index = collection.index_of(it);
+            return log_index.nth_block_position(block_num - log_data.first_block_num());
```

Comment:
It seems like, if I'm understanding this correctly, only a single partition is ever "loaded" at a time, and if a block is requested from a partition other then the one that is loaded, it closes the loaded partition and opens the needed partition.

For a node serving significant number of syncing peers that seems like it could result in a lot of thrashing. Tearing down a memory mapping and creating a new one is not cheap, and historically has involved actions like TLB flushes which need to halt all threads (I admit to not being up to date on the state of the art here, so maybe it's not as bad now as I remember).

But even forgetting the memory mapping behavior, I'm not sure how to resolve the thrashing of file descriptors: we can't keep them all open. A cache with simple LRU eviction? Or maybe just file descriptor thrashing isn't enough of a grind to worry about.

Because it appears we need to re-aquire resources, it seems very difficult to get the partitioned block log performance to be as good as the flat block log performance. A flat unpruned block log can serve multiple threads without any locking and with only a single resource acquisition at startup, for example.

Reply:
I am not exactly sure what you want me to do here. Increase or the number of cached file descriptor? Or we can wait until the PR is merged and then have some performance evaluation to decide whether it needs to be addressed?

Reply:
It's not necessarily an actionable item, just something to get everyone thinking about the performance and reliability side effects of this feature.

Can you describe what happens here if we fail to open the file (because, say, we've exhausted all file descriptors)? Clearly it `return {}` but then what happens? Will it get any peers in to a weird state or such?

Reply:
Not sure it is part of this effort, but would be nice to have a test that runs nodeos out of file descriptors and verifies it exits cleanly. Such a test would not necessarily run out of file descriptors right here which is why it likely should be a separate task.

Reply:
Also, maybe we should try to avoid this problem, by checking how many file descriptors we are allowed at startup, and not accepting connections if we get close to that number being in use.

Reply:
When it returns `{}`, it means the current nodeos instance does not have the requested block, hence the net plugin should be able to response appropriately. The current main brnach can handle this behavior when a nodeos does not have the full block log. This PR does not alter it at all.